libgadget/utils/system.h

```diff
@@ -83,4 +83,18 @@ int _MPIU_Barrier(const char * fn, const int ln, MPI_Comm comm);
 /* Fancy barrier which warns if there is a lot of imbalance. */
 #define MPIU_Barrier(comm) _MPIU_Barrier(__FILE__, __LINE__, comm)
 
+#ifdef DEBUG
+/* Checked versions of some MPI routines which make sure that the number we get out is sensible*/
+int MPI_Allreduce_Checked(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Op op, MPI_Comm comm, const int line, const char * file);
+int MPI_Reduce_Checked(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Op op, int root, MPI_Comm comm, const int line, const char * file);
+/* Defines, guarded so the implementation can still use the original*/
+#ifndef __UTILS_SYSTEM_C
```

Comment:
header that behaves differently if included from the impl is not a good pattern.

What about lifting the function rename macros to mpi-checked.h, and only include it in the soruce files that you want mpi to be checked?

Reply:
I agree! I really don't want a different header though - we have no other way of knowing that there is a bug here and I want maximal coverage, since Allreduce returning zero seems to only happen intermittently. It is too easy to forget the include if it is separate.